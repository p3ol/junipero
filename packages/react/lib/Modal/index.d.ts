import { ReactNode, MutableRefObject, ComponentPropsWithRef } from 'react';

export declare type ModalRef = {
  isJunipero: boolean;
  opened: boolean;
  close(): void;
  open(): void;
  toggle(): void;
  closeButtonRef: MutableRefObject<any>;
  contentRef: MutableRefObject<any>;
  innerRef: MutableRefObject<any>;
  wrapperRef: MutableRefObject<any>;
};

declare interface ModalProps extends ComponentPropsWithRef<any> {
  apparition?: string;
  children?: ReactNode | JSX.Element;
  className?: string;
  container?: string | Element | DocumentFragment;
  disabled?: boolean;
  opened?: boolean;
  closable?: boolean;
  animate?(
    modal: ReactNode | JSX.Element,
    options: { opened: boolean }
  ): void;
  onToggle?(props: { opened: boolean }): void;
  ref?: MutableRefObject<ModalRef | undefined>;
}

declare function Modal(props: ModalProps): ReactNode | JSX.Element;

export default Modal;
