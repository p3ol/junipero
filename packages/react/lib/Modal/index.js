import {
  forwardRef,
  useEffect,
  useImperativeHandle,
  useReducer,
  useRef,
} from 'react';
import { createPortal } from 'react-dom';
import PropTypes from 'prop-types';
import { classNames, mockState, ensureNode } from '@junipero/core';

import { Remove } from '../icons';

const Modal = forwardRef(({
  animate,
  children,
  className,
  apparition,
  container,
  disabled,
  opened,
  closable = true,
  onToggle,
  ...rest
}, ref) => {
  const innerRef = useRef();
  const contentRef = useRef();
  const wrapperRef = useRef();
  const closeButtonRef = useRef();
  const [state, dispatch] = useReducer(mockState, {
    opened: opened ?? false,
  });

  useImperativeHandle(ref, () => ({
    innerRef,
    contentRef,
    wrapperRef,
    closeButtonRef,
    opened: state.opened,
    open,
    close,
    toggle,
    isJunipero: true,
  }));

  useEffect(() => {
    dispatch({ opened: disabled ? false : !!opened });
  }, [disabled, opened]);

  const onBackdropClick = e => {
    if (
      disabled ||
      !closable ||
      (e.target !== wrapperRef.current && e.target !== closeButtonRef?.current)
    ) {
      return;
    }

    e.preventDefault();
    close();
  };

  const open = () => {
    if (disabled) {
      return;
    }

    dispatch({ opened: true });
    onToggle?.({ opened: true });
  };

  const close = () => {
    if (disabled) {
      return;
    }

    dispatch({ opened: false });
    onToggle?.({ opened: false });
  };

  const toggle = () => {
    if (disabled) {
      return;
    }

    state.opened = !state.opened;
    dispatch({ opened: state.opened });
    onToggle?.({ opened: state.opened });
  };

  const Wrapper = ({ transitionState }) => (
    <div
      { ...rest }
      ref={innerRef}
      className={classNames(
        'junipero',
        'modal',
        {
          opened: state.opened,
        },
        className,
      )}
    >
      <div
        tabIndex={0}
        onClick={onBackdropClick}
        ref={wrapperRef}
        className="wrapper"
      >
        <div ref={contentRef} className="content">
          { closable && (
            <a
              href="#"
              ref={closeButtonRef}
              className="close"
              onClick={onBackdropClick}
            >
              <Remove />
            </a>
          ) }
          {(!transitionState || ['entered', 'exiting'].find(
            state => state === transitionState)
          ) &&
            children
          }
        </div>
      </div>
    </div>
  );
  Wrapper.propTypes = {
    transitionState: PropTypes.string,
  };
  const content = animate
    ? animate(Wrapper, { opened: state.opened }) : <Wrapper />;

  return state.opened || animate || apparition === 'css'
    ? container ? createPortal(content, ensureNode(container)) : content
    : null;
});

Modal.displayName = 'Modal';
Modal.propTypes = {
  animate: PropTypes.func,
  apparition: PropTypes.string,
  container: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
    PropTypes.object,
    PropTypes.func,
  ]),
  closable: PropTypes.bool,
  disabled: PropTypes.bool,
  onToggle: PropTypes.func,
  opened: PropTypes.bool,
};

export default Modal;
