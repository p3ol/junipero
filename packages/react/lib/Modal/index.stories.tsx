import { useRef } from 'react';
import { action } from 'storybook/actions';
import { appearBounceModal, slideInLeftModal } from '@junipero/transitions';

import Button from '../Button';
import Modal, { ModalRef } from './index';

export default { title: 'react/Modal' };

export const basic = () => {
  const modalRef = useRef<ModalRef>(null);

  return (
    <>
      <Button className="primary" onClick={() => modalRef.current?.open()}>
        Open me
      </Button>
      <Modal ref={modalRef} onToggle={action('toggle')}>Test</Modal>
    </>
  );
};

export const animated = () => {
  const modalRef = useRef<ModalRef>(null);

  return (
    <>
      <Button className="primary" onClick={() => modalRef.current?.open()}>
        Open me
      </Button>
      <Modal ref={modalRef} animate={appearBounceModal}>
        Test
      </Modal>
    </>
  );
};

export const panel = () => {
  const modalRef = useRef<ModalRef>(null);

  return (
    <>
      <Button className="primary" onClick={() => modalRef.current?.open()}>
        Open me
      </Button>
      <Modal
        className="panel"
        ref={modalRef}
        animate={slideInLeftModal}
        onToggle={action('toggle')}
      >
        Test
      </Modal>
    </>
  );
};

export const nonClosablePanel = () => {
  const modalRef = useRef<ModalRef>(null);

  return (
    <>
      <Button className="primary" onClick={() => modalRef.current?.open()}>
        Open me
      </Button>
      <Modal
        closable={false}
        className="panel"
        ref={modalRef}
        onToggle={action('toggle')}
      >
        <div>Test</div>
        <div>
          <Button onClick={() => modalRef.current?.close()}>Close me</Button>
        </div>
      </Modal>
    </>
  );
};
