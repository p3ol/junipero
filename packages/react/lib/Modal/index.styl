@require '../theme/colors'

.junipero.modal
  --background-color: rgba($color-midnight, .9)

  position: fixed
  top: 0
  right: 0
  bottom: 0
  left: 0
  overflow: auto
  display: block

  &.apparition-css
    display: none

    &.opened
      display: block

  & > .wrapper
    width: 100%
    height: 100%
    position: absolute
    z-index: 2
    padding: 50px
    background: var(--background-color)
    overflow: auto
    display: flex
    align-items: center
    justify-content: center

    & > .content
      width: 700px
      max-width: 100%
      background: $color-white
      min-height: 500px
      position: relative
      z-index: 3
      border: none
      border-radius: 4px
      margin: 0 auto
      padding: 20px

      & > a.close
        display: inline-block
        width: 15px
        height: 15px
        position: absolute
        cursor: pointer
        top: 10px
        right: 10px

        &:before, &:after
          content: ''
          height: 100%
          position: absolute
          width: 2px
          background: $color-shuttle-gray
          left: calc(50% - 1px)
          top: 0

        &:before
          transform: rotate(45deg)

        &:after
          transform: rotate(-45deg)

  &.panel
    & > .wrapper
      padding: 0
      justify-content: flex-end
      overflow: hidden

      & > .content
        height: 100%
        overflow: auto
        margin: 0
        border-radius: 0
