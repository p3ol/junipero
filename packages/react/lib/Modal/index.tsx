import {
  ComponentPropsWithRef,
  MutableRefObject,
  ReactNode,
  forwardRef,
  useEffect,
  useImperativeHandle,
  useReducer,
  useRef,
} from 'react';
import { createPortal } from 'react-dom';
import PropTypes from 'prop-types';
import { classNames, mockState, ensureNode } from '@junipero/core';

import { Remove } from '../icons';
import { ForwardedProps } from '../utils';

export declare type ModalRef = {
  isJunipero: boolean;
  opened: boolean;
  close(): void;
  open(): void;
  toggle(): void;
  closeButtonRef: MutableRefObject<any>;
  contentRef: MutableRefObject<any>;
  innerRef: MutableRefObject<any>;
  wrapperRef: MutableRefObject<any>;
};

declare interface ModalProps extends ComponentPropsWithRef<any> {
  apparition?: string;
  children?: ReactNode | JSX.Element;
  className?: string;
  container?: Element | DocumentFragment;
  disabled?: boolean;
  opened?: boolean;
  closable?: boolean;
  animate?(
    modal: ReactNode | JSX.Element,
    options: { opened: boolean, onExited: () => void }
  ): JSX.Element;
  onToggle?(props: { opened: boolean }): void;
  ref?: MutableRefObject<ModalRef | undefined>;
}

const Modal = forwardRef(({
  animate,
  children,
  className,
  apparition,
  container,
  disabled,
  opened,
  closable = true,
  onToggle,
  ...rest
}: ModalProps, ref) => {
  const innerRef = useRef();
  const contentRef = useRef();
  const wrapperRef = useRef();
  const closeButtonRef = useRef();
  const [state, dispatch] = useReducer(mockState, {
    opened: opened ?? false,
    visible: opened ?? false,
  });

  useImperativeHandle(ref, () => ({
    innerRef,
    contentRef,
    wrapperRef,
    closeButtonRef,
    opened: state.opened,
    open,
    close,
    toggle,
    isJunipero: true,
  }));

  useEffect(() => {
    dispatch({ opened: disabled ? false : !!opened });
  }, [disabled, opened]);

  const onBackdropClick = e => {
    if (
      disabled ||
      !closable ||
      (e.target !== wrapperRef.current && e.target !== closeButtonRef?.current)
    ) {
      return;
    }

    e.preventDefault();
    close();
  };

  const onCloseClick = e => {
    e.preventDefault();

    if (disabled || !closable) {
      return;
    }

    close();
  };

  const open = () => {
    if (disabled) {
      return;
    }

    dispatch({ opened: true, visible: true });
    onToggle?.({ opened: true });
  };

  const close = () => {
    if (disabled) {
      return;
    }

    dispatch({ opened: false });
    onToggle?.({ opened: false });
  };

  const toggle = () => {
    if (disabled) {
      return;
    }

    if (state.opened) {
      close();
    } else {
      open();
    }
  };

  const onAnimationExit = () => {
    dispatch({ visible: false });
  };

  const wrapper = (
    <div
      { ...rest }
      ref={innerRef}
      className={classNames(
        'junipero',
        'modal',
        {
          opened: state.opened,
        },
        className,
      )}
    >
      <div
        tabIndex={0}
        onClick={onBackdropClick}
        ref={wrapperRef}
        className="wrapper"
      >
        <div ref={contentRef} className="content">
          { closable && (
            <a
              href="#"
              ref={closeButtonRef}
              className="close"
              onClick={onCloseClick}
            >
              <Remove />
            </a>
          ) }
          { children }
        </div>
      </div>
    </div>
  );

  const content = animate
    ? animate(wrapper, {
      opened: state.opened,
      onExited: onAnimationExit,
    }) : wrapper;

  return state.opened || (animate && state.visible) || apparition === 'css'
    ? container ? createPortal(content, container) : content
    : null;
}) as ForwardedProps<ModalProps, ModalRef>;

Modal.displayName = 'Modal';
Modal.propTypes = {
  animate: PropTypes.func,
  apparition: PropTypes.string,
  container: PropTypes.oneOfType([
    PropTypes.any, //TODO find a way to check if it's a valid DOM node
  ]),
  closable: PropTypes.bool,
  disabled: PropTypes.bool,
  onToggle: PropTypes.func,
  opened: PropTypes.bool,
};

export default Modal;
