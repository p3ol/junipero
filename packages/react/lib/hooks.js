import { useContext } from 'react';

import { FieldControlContext } from './contexts';

export const useFieldControl = () => useContext(FieldControlContext);
