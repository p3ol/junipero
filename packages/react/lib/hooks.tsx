import { useContext } from 'react';

import {
  AlertsContextType,
  AlertsContext,
  DropdownContext,
  DropdownContextType,
  FieldContextType,
  FieldControlContext,
  ListContext,
  ListContextType,
  ToastsContext,
  ToastsContextType,
  ModalContext,
  ModalContextType,
} from './contexts';

export const useFieldControl =
  () => useContext<FieldContextType>(FieldControlContext);
export const useDropdown = () =>
  useContext<DropdownContextType>(DropdownContext);
export const useAlerts = () => useContext<AlertsContextType>(AlertsContext);
export const useList = () => useContext<ListContextType>(ListContext);
export const useToasts = () => useContext<ToastsContextType>(ToastsContext);
export const useModal = () => useContext<ModalContextType>(ModalContext);

export { AlertsContext, DropdownContext, FieldControlContext, ListContext, ToastsContext, ModalContext };
