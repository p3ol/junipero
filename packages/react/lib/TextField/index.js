import {
  forwardRef,
  useEffect,
  useImperativeHandle,
  useReducer,
  useRef,
} from 'react';
import { classNames, mockState, exists } from '@junipero/core';
import PropTypes from 'prop-types';

import { useFieldControl } from '../hooks';

const TextField = forwardRef(({
  min,
  max,
  autoFocus,
  children,
  className,
  valid,
  value,
  disabled = false,
  required = false,
  type = 'text',
  onBlur,
  onChange,
  onFocus,
  onWheel,
  onValidate = (val, { required }) => !!val || !required,
  ...rest
}, ref) => {
  const innerRef = useRef();
  const inputRef = useRef();
  const { update: updateControl } = useFieldControl();
  const [state, dispatch] = useReducer(mockState, {
    value: value ?? '',
    valid: valid ?? false,
    focused: autoFocus ?? false,
    dirty: false,
  });

  useImperativeHandle(ref, () => ({
    innerRef,
    inputRef,
    value: state.value,
    valid: state.valid,
    dirty: state.dirty,
    focused: state.focused,
    focus,
    blur,
    reset,
    setDirty,
    isJunipero: true,
  }));

  useEffect(() => {
    if (exists(value)) {
      state.value = value;
      state.valid = onValidate(state.value, { dirty: state.dirty, required });
      dispatch({ value: state.value, valid: state.valid });
      updateControl?.({ valid: state.valid, dirty: state.dirty });
    }
  }, [value]);

  useEffect(() => {
    dispatch({ valid: valid ?? false });
    updateControl?.({ valid: state.valid });
  }, [valid]);

  const parseValue = value => {
    const value_ = value.replace(',', '.');
    const parsedValue = parseFloat(value_);

    return !isNaN(parsedValue) ? parsedValue : value_;
  };

  const onChange_ = e => {
    /* html input disabled attribute will prevent onChange if present anyway */
    /* istanbul ignore next: just in case */
    if (disabled) {
      return;
    }

    if (type === 'number') {
      value = parseValue(e?.target?.value);
    } else {
      value = e?.target?.value;
    }

    state.value = value;
    state.dirty = true;
    state.valid = onValidate(
      state.value, { dirty: state.dirty, required }
    );
    dispatch({ value: state.value, dirty: state.dirty });
    onChange?.({ value: state.value, valid: state.valid, dirty: state.dirty });
    updateControl?.({ dirty: true, valid: state.valid });
  };

  const onFocus_ = e => {
    dispatch({ focused: true });
    updateControl?.({ focused: true });
    onFocus?.(e);
  };

  const onBlur_ = e => {
    dispatch({ focused: false });
    updateControl?.({ focused: false });
    onBlur?.(e);
  };

  const focus = () => {
    inputRef.current?.focus();
  };

  const blur = () => {
    inputRef.current?.blur();
  };

  const reset = () => {
    dispatch({ value: value ?? '', valid: valid ?? false, dirty: false });
    updateControl?.({ dirty: false, valid: valid ?? false });
  };

  const isEmpty = () =>
    !exists(state.value) || state.value === '';

  const setDirty = dirty => {
    state.dirty = dirty;
    dispatch({ dirty });
    updateControl?.({ dirty });
  };

  const Tag = rest.rows > 1 ? 'textarea' : 'input';

  return (
    <div
      ref={innerRef}
      className={classNames(
        'junipero',
        'text-field',
        state.dirty ? 'dirty' : 'pristine',
        !state.valid && state.dirty ? 'invalid' : 'valid',
        { empty: isEmpty(), focused: state.focused, disabled },
        className
      )}
    >
      <Tag
        { ...rest }
        min={min}
        max={max}
        autoFocus={autoFocus}
        className="field"
        ref={inputRef}
        disabled={disabled}
        required={required}
        type={Tag === 'textarea' ? undefined : type}
        value={state.value}
        onChange={onChange_}
        onFocus={onFocus_}
        onBlur={onBlur_}
        onWheel={onWheel ?? (() => inputRef.current?.blur())}
      />
      { children }
    </div>
  );
});

TextField.displayName = 'TextField';
TextField.propTypes = {
  min: PropTypes.number,
  max: PropTypes.number,
  autoFocus: PropTypes.bool,
  disabled: PropTypes.bool,
  required: PropTypes.bool,
  type: PropTypes.string,
  valid: PropTypes.bool,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  onFocus: PropTypes.func,
  onWheel: PropTypes.func,
  onValidate: PropTypes.func,
};

export default TextField;
