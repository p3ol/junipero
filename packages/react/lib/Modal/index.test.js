import { createRef } from 'react';
import { render, fireEvent, act } from '@testing-library/react';

import Modal from './';

describe('<Modal />', () => {
  it('should render', () => {
    const ref = createRef();
    const { container, unmount } = render(
      <div className="container">
        <Modal opened={true} container=".container" ref={ref} />
      </div>
    );
    expect(container).toMatchSnapshot();
    unmount();
  });

  it('should remove modal from DOM when using close method', async () => {
    const ref = createRef();
    const { container, unmount } = render(<Modal ref={ref} />);
    await act(async () => { ref.current.open(); });
    expect(container).toMatchSnapshot('opened');
    await act(async () => { ref.current.close(); });
    expect(container).toMatchSnapshot('closed');
    unmount();
  });

  it('should not toggle modal if it is disabled', async () => {
    const ref = createRef();
    const { container, unmount } = render(
      <Modal ref={ref} disabled={true} />
    );
    await act(async () => { ref.current.open(); });
    expect(container).toMatchSnapshot();
    unmount();
  });

  it('should close modal when clicking on backdrop', async () => {
    const ref = createRef();
    const { container, unmount } = render(
      <div className="container">
        <Modal container=".container" ref={ref} />
      </div>
    );
    await act(async () => { ref.current.open(); });
    expect(container).toMatchSnapshot('opened');
    fireEvent.click(container.querySelector('.wrapper'));
    expect(container).toMatchSnapshot('closed');
    unmount();
  });

  it('should not close modal if clicked inside', async () => {
    const ref = createRef();
    const { container, unmount } = render(
      <div className="container">
        <Modal container=".container" ref={ref} />
      </div>
    );
    await act(async () => { ref.current.open(); });
    expect(container).toMatchSnapshot();
    fireEvent.click(container.querySelector('.content'));
    expect(container).toMatchSnapshot();
    unmount();
  });

  it('should not close modal closable = false', async () => {
    const ref = createRef();
    const { container, unmount } = render(
      <div className="container">
        <Modal container=".container" closable={false} ref={ref} />
      </div>
    );
    await act(async () => { ref.current.open(); });
    expect(container).toMatchSnapshot();
    fireEvent.click(container.querySelector('.wrapper'));
    expect(container).toMatchSnapshot();
    unmount();
  });
});
