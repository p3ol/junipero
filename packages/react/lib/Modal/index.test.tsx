import { createRef } from 'react';
import { render, fireEvent, act } from '@testing-library/react';
import { appearBounceModal } from '@junipero/transitions';

import Modal, { ModalRef } from '.';

describe('<Modal />', () => {
  it('should render', () => {
    const ref = createRef<ModalRef>();
    const { container, unmount } = render(
      <div className="container">
        <Modal opened={true} container=".container" ref={ref} />
      </div>
    );
    expect(container).toMatchSnapshot();
    unmount();
  });

  it('should remove modal from DOM when using close method', async () => {
    const ref = createRef<ModalRef>();
    const { container, unmount } = render(<Modal ref={ref} />);
    await act(async () => Promise.resolve(ref.current.open()));
    expect(container).toMatchSnapshot('opened');
    await act(async () => Promise.resolve(ref.current.close()));
    expect(container).toMatchSnapshot('closed');
    unmount();
  });

  it('should not toggle modal if it is disabled', async () => {
    const ref = createRef<ModalRef>();
    const { container, unmount } = render(
      <Modal ref={ref} disabled={true} />
    );
    await act(async () => Promise.resolve(ref.current.open()));
    expect(container).toMatchSnapshot();
    unmount();
  });

  it('should close modal when clicking on backdrop', async () => {
    const ref = createRef<ModalRef>();
    const { container, unmount } = render(
      <div className="container">
        <Modal container=".container" ref={ref} />
      </div>
    );
    await act(async () => Promise.resolve(ref.current.open()));
    expect(container).toMatchSnapshot('opened');
    fireEvent.click(container.querySelector('.wrapper'));
    expect(container).toMatchSnapshot('closed');
    unmount();
  });

  it('should not close modal if clicked inside', async () => {
    const ref = createRef<ModalRef>();
    const { container, unmount } = render(
      <div className="container">
        <Modal container=".container" ref={ref} />
      </div>
    );
    await act(async () => Promise.resolve(ref.current.open()));
    expect(container).toMatchSnapshot();
    fireEvent.click(container.querySelector('.content'));
    expect(container).toMatchSnapshot();
    unmount();
  });

  it('should not close modal closable = false', async () => {
    const ref = createRef<ModalRef>();
    const { container, unmount } = render(
      <div className="container">
        <Modal container=".container" closable={false} ref={ref} />
      </div>
    );
    await act(async () => Promise.resolve(ref.current.open()));
    expect(container).toMatchSnapshot();
    fireEvent.click(container.querySelector('.wrapper'));
    expect(container).toMatchSnapshot();
    unmount();
  });

  it('should allow to animate modal', () => {
    jest.useFakeTimers();
    const ref = createRef<ModalRef>();
    const { container, unmount } = render(
      <div className="container">
        <Modal animate={appearBounceModal} ref={ref} />
      </div>
    );
    act(() => {
      ref.current.open();
      jest.advanceTimersByTime(100);
    });
    expect(container).toMatchSnapshot('opened');
    act(() => {
      ref.current.close();
      jest.advanceTimersByTime(100);
    });
    expect(container).toMatchSnapshot('closed');

    jest.useRealTimers();
    unmount();
  });
});
