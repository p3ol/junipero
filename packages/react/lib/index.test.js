import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import TextField from './TextField';

describe('mockState: inside react', () => {
  it('should fail update the state correctly and return the old ' +
    'state', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch({ value: { ...state.value, content: ['foo'] } });
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch({ value: { ...state.value, text: value } });
      };

      const onContentChange = () => {
        dispatch({
          value: { ...state.value, content: [...state.value.content, 'bar'] },
        });
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content but with old text value');

    unmount();
  });

  it('should not fail update the state correctly and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
