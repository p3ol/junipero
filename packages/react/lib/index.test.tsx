import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import type { FieldContent } from './types';
import TextField from './TextField';

interface State {
  value: {
    text: string;
    content: string[];
  };
}

describe('mockState: inside react', () => {
  it('should correctly update the state and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState<State>, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }: FieldContent) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
