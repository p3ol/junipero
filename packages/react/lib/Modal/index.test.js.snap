// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal /> should close modal when clicking on backdrop: closed 1`] = `
<div>
  <div
    class="container"
  />
</div>
`;

exports[`<Modal /> should close modal when clicking on backdrop: opened 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="junipero modal opened"
    >
      <div
        class="wrapper"
        tabindex="0"
      >
        <div
          class="content"
        >
          <a
            class="close"
            href="#"
          >
            <svg
              class="junipero icon remove"
              height="10"
              viewBox="0 0 9 10"
              width="10"
            >
              <path
                d="M8 1.5L1 8.5"
              />
              <path
                d="M1 1.5L8 8.5"
              />
            </svg>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Modal /> should not close modal closable = false 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="junipero modal opened"
    >
      <div
        class="wrapper"
        tabindex="0"
      >
        <div
          class="content"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<Modal /> should not close modal closable = false 2`] = `
<div>
  <div
    class="container"
  >
    <div
      class="junipero modal opened"
    >
      <div
        class="wrapper"
        tabindex="0"
      >
        <div
          class="content"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<Modal /> should not close modal if clicked inside 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="junipero modal opened"
    >
      <div
        class="wrapper"
        tabindex="0"
      >
        <div
          class="content"
        >
          <a
            class="close"
            href="#"
          >
            <svg
              class="junipero icon remove"
              height="10"
              viewBox="0 0 9 10"
              width="10"
            >
              <path
                d="M8 1.5L1 8.5"
              />
              <path
                d="M1 1.5L8 8.5"
              />
            </svg>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Modal /> should not close modal if clicked inside 2`] = `
<div>
  <div
    class="container"
  >
    <div
      class="junipero modal opened"
    >
      <div
        class="wrapper"
        tabindex="0"
      >
        <div
          class="content"
        >
          <a
            class="close"
            href="#"
          >
            <svg
              class="junipero icon remove"
              height="10"
              viewBox="0 0 9 10"
              width="10"
            >
              <path
                d="M8 1.5L1 8.5"
              />
              <path
                d="M1 1.5L8 8.5"
              />
            </svg>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Modal /> should not toggle modal if it is disabled 1`] = `<div />`;

exports[`<Modal /> should remove modal from DOM when using close method: closed 1`] = `<div />`;

exports[`<Modal /> should remove modal from DOM when using close method: opened 1`] = `
<div>
  <div
    class="junipero modal opened"
  >
    <div
      class="wrapper"
      tabindex="0"
    >
      <div
        class="content"
      >
        <a
          class="close"
          href="#"
        >
          <svg
            class="junipero icon remove"
            height="10"
            viewBox="0 0 9 10"
            width="10"
          >
            <path
              d="M8 1.5L1 8.5"
            />
            <path
              d="M1 1.5L8 8.5"
            />
          </svg>
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Modal /> should render 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="junipero modal opened"
    >
      <div
        class="wrapper"
        tabindex="0"
      >
        <div
          class="content"
        >
          <a
            class="close"
            href="#"
          >
            <svg
              class="junipero icon remove"
              height="10"
              viewBox="0 0 9 10"
              width="10"
            >
              <path
                d="M8 1.5L1 8.5"
              />
              <path
                d="M1 1.5L8 8.5"
              />
            </svg>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
